draft-specifications/Model.md

```diff
@@ -0,0 +1,549 @@
+- Every finite Float64 is equal to any other Float64 with the same bitwise
+  representation.
+- Every infinite Float64 is equal to any other infinite Float64 of the same
+  sign.
+- There is only one NaN value and it is equal to itself.
+
+> JavaScript's `Object.is` is consistent with Pass Invariant Equality, whereas
+> `==` and `===` are not.
+
+## String
+
+([JSON](#json-invariants)†)
+
+A string of Unicode code points excluding lone (or unpaired) surrogates
+(U+D800-U+DFFF).
+Strings are distinguished from [Selector](#selector) s by type, not content.
+
+> - **Guile**: `""`
+> - **JavaScript**: `''`
+> - **Python**: `''`
+>
+> †Strings participate in the [JSON subset](#json-invariants) of OCapN except
+> any strings that contain unpaired/lone surrogates.
+>
+> A string's content must be expressible in UTF-8.
+> Some two-byte Unicode encodings, as in JavaScript strings, can contain
+> unpaired/lone surrogates that have no valid expression in any UTF and so
+> cannot be carried by OCapN.
+>
+> Notes: [January 2024 meeting
+> notes](https://github.com/ocapn/ocapn/blob/main/meeting-minutes/2024-01-09.md)
+> record that we agreed that strings can only be well-formed Unicode, i.e.,
+> cannot contain unpaired surrogates.
+> For JavaScript, if a string does not pass [the `isWellFormed`
+> predicate](https://github.com/tc39/proposal-is-usv-string), then it is not a
+> Passable string.
+
+For purposes of [Pass Invariant Equality](#pass-invariant-equality), a pair of
+Strings are equal if they have the same quantity of Unicode code points and
+have the same respective Unicode code points in order.
+
+## ByteArray
+
+An array of bytes.
+
+> - **Guile**: `#vu8()`
+> - **JavaScript**: `new ArrayBuffer()`
+> - **Python**: `b''`
+>
+> Tracking: https://github.com/ocapn/ocapn/issues/48
+>
+> At the [November 14, 2023](https://github.com/ocapn/ocapn/issues/94) meeting,
+> we agreed to settle on the prefix "byte" and agreed to decide at the next
+> meeting based on a [reactji
+> poll](https://github.com/ocapn/ocapn/issues/48#issuecomment-1811097196).
+>
+> [January 2024 meeting
+> notes](https://github.com/ocapn/ocapn/blob/main/meeting-minutes/2024-01-09.md)
+> record that we agreed on ByteArray because it was the winner of the poll, and
+> we had already agreed to resolve this specific issue by poll among these
+> three choices.
+>
+> The JavaScript representation of a ByteArray is an `ArrayBuffer` which may
+> be made immutable with the proposed JavaScript [Immutable
+> ArrayBuffer](https://github.com/tc39/proposal-immutable-arraybuffer) feature.
+
+For purposes of [Pass Invariant Equality](#pass-invariant-equality), a pair of
+ByteArrays are equal if they have the same quantity of bytes and have the same
+respective bytes in order.
+
+## Selector
+
+A string of Unicode code points excluding lone (or unpaired) surrogates
+(U+D800-U+DFFF).
+Selectors are distinguished from [String](#string)s by type, not content.
```

Comment:
```suggestion
A sequence of Unicode code points excluding surrogates (U+D800-U+DFFF).
An implementation of OCapN in JavaScript or another language in which
strings are based on UTF-16 may include 16-bit surrogate code _units_
(0xD800-0xDFFF), but only as part of a surrogate pair which encodes a
single code point greater than or equal to U+10000 (i.e., each OCapN
string can be encoded into a well-formed Unicode string in any encoding
form).

Selectors are distinguished from [String](#string)s by type, not content.
```

Reply:
I’ve doctored this into the non-normative qualifier below the example and synced it with selector instead.

```
> A string/selector's content must be expressible in UTF-8.
> Some two-byte Unicode encodings, as in JavaScript strings, can contain
> 16-bit surrogate code _units_ in the range from 0xD800-0xDFFF.
> Pairs of surrogate code units correspond to a single Unicode code _point_
> greater than or equal to U-10000 and can be expressed in UTF-8.
> However unpaired or lone surrogates have no valid expression in any UTF
> and so cannot be carried by OCapN.

```